### ISSUE TITLE
Testing - Appveyor, MinGW, trunk, minitest

### ISSUE THREAD

**Issue by MSP-Greg** [rubygems/rubygems#1984]
Recently ran 'rake test' with a trunk MinGW build (ruby 2.5.0dev (2017-08-03 trunk 59487) [x64-mingw32]).

Summary was:
```
1950 runs, 5715 assertions, 2 failures, 0 errors, 35 skips
```

Two failures were:
```
  1) Failure:
TestGemResolver#test_sorts_by_source_then_version [E:/GitHub/rubygems/test/rubygems/test_gem_resolver.rb:709]:
Set of gems was not the same: ["some-dep-0.0.2"] != ["some-dep-1.0.0"].
--- expected
+++ actual
@@ -1,6 +1,6 @@
 [Gem::Specification.new do |s|
    s.name = "some-dep"
-   s.version = Gem::Version.new("0.0.2")
+   s.version = Gem::Version.new("1.0.0")
    s.installed_by_version = Gem::Version.new("0")
    s.authors = ["A User"]
    s.date = Time.utc(2017, 8, 5)

 22) Failure:
TestGem#test_use_gemdeps_missing_gem [E:/GitHub/rubygems/test/rubygems/test_gem.rb:1756]:
In stderr.
--- expected
+++ actual
@@ -1,4 +1,4 @@
-"Could not find gem 'a x86-mswin32' in any of the gem sources listed in your Gemfile.
+"Could not find gem 'a x86-mswin32' in any of the gem sources listed in your Gemfile or available on this machine.
 " +
 "You may need to `gem install -g` to install missing gems
 " +

```

**Setup**

A. Replaced `~>` constraint with `>=` for builder, minitest, rake,and rdoc.

B. Applied [PR 1981](https://github.com/rubygems/rubygems/pull/1981).

C. Set path with MSYS2/MINGW64 bin folders, added GIT env variable with 'no space' path.

**Questions**

1. Current version of [minitest](https://rubygems.org/gems/minitest) seems to show no version constraint on ruby.  Is there any interest in changing the version constraint?  Using the new version, I had several of the following two messages:

```
MiniTest::Unit::TestCase is now Minitest::Test
DEPRECATED: Use assert_nil if expecting nil
```

2. I am not familiar with the internals of RubyGems.  If anyone has any 'tips' for tracking down the two failures...  The second failure looks like it is due to incorrect string matching.

3. Appveyor - current script only tests thru ruby 2.2.  I'm willing to look at adding to that, but I'd like to do everything at once, especially the minitest update.  I'm keeping a current trunk build (with OpenSSL 1.1.0) on BinTray, so it could also be added, using [here](https://ci.appveyor.com/project/MSP-Greg/appveyor-ruby), trunk build info is at the bottom.

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
We can't use the latest minitest because it's incompatible with ruby trunk's testing framework.

**Comment by MSP-Greg** [comment]
```
class Gem::TestCase < (defined?(Minitest::Test) ? Minitest::Test : MiniTest::Unit::TestCase)
```

**Comment by MSP-Greg** [comment]
Ran a trunk build this morning with all applicable patches (used for `rake test` in my fork) applied to ruby trunk.  All okay, test logs in BinTray file [zlogs_trunk_2017-08-06_59522.7z](https://dl.bintray.com/msp-greg/ruby_windows/:zlogs_trunk_2017-08-06_59522.7z)

One additional failure (`TestGemLocalRemoteOptions#test_clear_sources_option_idiot_proof `) is intermittent.  As a final check, I'll patch appveyor.yml and push my fork to travis and appveyor...

`rake test` output (excluding requires):
```
E:\>cd E:\GitHub\rubygems & set GIT=E:/GitHub/ruby-loco/git/cmd/git.exe & rake -rdevkit test

Run options: --seed 51526

# Running:

................................................................................................................................S.....................
......................................................................................................................................................
.............................................................S..............................................................................S.........
........................................S.......................S.............S....................................S.....S...................S........
...........................S......S...........................S.......................................................................................
.......................................................................................S..............................................................
..............................................................................................................................F.......................
..........................................................................................................................S..........................S
.........................S.........................................................SSS................................................................
.........F......................................S.SS.S.....SSSS.................................S..S..................................................
........................................................................................................S....S........................................
........................S.............................................................................................................................
........................................................................................S.....S.........S.............................................


Finished in 193.002085s, 10.1035 runs/s, 29.6163 assertions/s.

  1) Failure:
TestGemLocalRemoteOptions#test_clear_sources_option_idiot_proof [E:/GitHub/rubygems/test/rubygems/test_gem_local_remote_options.rb:48]:
--- expected
+++ actual
@@ -1 +1,3 @@
-["https://rubygems.org/"]
+#<Gem::SourceList:0xXXXXXX
+ @sources=
+  [[Remote: http://test.example.com/ API URI: http://test.example.com/]]>


  2) Failure:
TestGemResolver#test_sorts_by_source_then_version [E:/GitHub/rubygems/test/rubygems/test_gem_resolver.rb:709]:
Set of gems was not the same: ["some-dep-0.0.2"] != ["some-dep-1.0.0"].
--- expected
+++ actual
@@ -1,6 +1,6 @@
 [Gem::Specification.new do |s|
    s.name = "some-dep"
-   s.version = Gem::Version.new("0.0.2")
+   s.version = Gem::Version.new("1.0.0")
    s.installed_by_version = Gem::Version.new("0")
    s.authors = ["A User"]
    s.date = Time.utc(2017, 8, 6)


1950 runs, 5716 assertions, 2 failures, 0 errors, 35 skips

You have skipped tests. Run with --verbose for details.
rake aborted!
Command failed with status (1): [C:/ruby25_64/bin/ruby.exe -w -Ilib;bin;tes...]

Tasks: TOP => test
(See full trace by running task with --trace)
```

**Comment by MSP-Greg** [comment]
I've got results for three 'builds' on [Appveyor](https://ci.appveyor.com/project/MSP-Greg/rubygems/history).

Changes:

1. Currently, tests are run on Ruby 1.9.3 thru 2.2.6.  I added 64 bit builds of 2.3.3, 2.4.1, and trunk.

2. Minitest was installed if the installed version was less that 5.8.

3. Build gem was installed to allow 'Gem::Indexer tests', and ENV['GIT'] was set correctly.

4. [PR 1981](https://github.com/rubygems/rubygems/pull/1981) is included.

5. Versions 1.9.3 thru 2.3.3 intermittently passed.

6. 2.4.1 and trunk always failed, and failures always included `TestGemResolver#test_sorts_by_source_then_version`.

7. Set Appveyor cache to only update current job's cache, as opposed to updating every cache on every job.
